torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+import os
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+import textwrap
+
+import torch
+
+
+class _ValgrindWrapper(object):
+    def __init__(self):
+        self._valgrind_available = not subprocess.run(
+            ["which", "valgrind"], capture_output=True).returncode
+
+        self._build_type = None
+        build_search = re.search("BUILD_TYPE=(.+),", torch.__config__.show())
+        if build_search is not None:
+            self._build_type = build_search.groups()[0].split(",")[0]
+
+        self._baseline_cache = {}
+
+    def _validate(self):
+        if not self._valgrind_available:
+            raise OSError("Could not find `valgrind`")
+
+        if self._build_type is None:
+            raise OSError("Could not determine BUILD_TYPE")
+
+        if self._build_type != "RelWithDebInfo":
+            raise OSError(
+                "PyTorch sould be built with REL_WITH_DEB_INFO=1 "
+                "(debug symbols) in order to produce a useful C++ call graph, "
+                f"got BUILD_TYPE={self._build_type} instead"
+            )
+
+    def collect_callgrind(self, stmt: str, setup: str, number: int, num_threads: int):
+        self._validate()
+        if number not in self._baseline_cache:
+            self._baseline_cache[(number, num_threads)] = self._invoke(
+                stmt="pass", setup="pass", number=number, num_threads=num_threads)
+
+        # We include `pass` so it matches the baseline and reduces the instruction diff.
+        fn_counts = self._invoke(
+            stmt=f"pass;{stmt}", setup=setup, number=number, num_threads=num_threads)
+
+        count_dict = {fn: count for count, fn in fn_counts}
+        for count, fn in self._baseline_cache[(number, num_threads)]:
+            count_dict.setdefault(fn, 0)
+            count_dict[fn] -= count
+            if not count_dict[fn]:
+                count_dict.pop(fn)
+
+        fn_diff_counts = []
+        for fn, count in sorted(count_dict.items(), key=lambda x: x[1], reverse=True):
+            if count < 0:
+                # The profiling process is not 100% deterministic. For instance
+                # the subprocess calls to control callgrind poll resulting in
+                # a non-deterministic number of instructions. This delta is
+                # generally small.
+                # FIXME: Validate that differences are small and expected, and warn otherwise.
+                break
+            fn_diff_counts.append((count, fn))
+        return fn_diff_counts
+
+    def _invoke(self, stmt: str, setup: str, number: int, num_threads: int):
+        working_dir = tempfile.mkdtemp()
+
+        _, script_file = tempfile.mkstemp(prefix="timer_callgrind_", suffix=".py", dir=working_dir)
```

Comment:
technically those don't need to be random files because the directory is already random?

Reply:
Done.